File: Programmers/Level 2/Level2_할인행사.py
want = ["banana", "apple", "rice", "pork", "pot"]
number = [3, 2, 2, 2, 1]
discount = ["chicken", "apple", "apple", "banana", "rice", "apple", "pork", "banana", "pork", "rice", "pot",
             "banana", "apple", "banana"]

def solution(want, number, discount):
    answer = 0
    want_dictionary = dict(zip(want, number))
    print(want_dictionary)

    for i in range(len(discount)):
        if len(discount[i:i+10]) < sum(number):
            break

        print("{}일차 = / PRODUCT : ".format(i + 1), end=' ')
        isAllFind = True
        for product in want:
            if discount[i:i+10].count(product) < want_dictionary.get(product):
                isAllFind = False
                break
            print("{}, {}".format(product, want_dictionary.get(product)), end=' ')
        print()

        if isAllFind == True:
            answer += 1

    return answer

print(solution(want, number, discount))

print(discount.count("apple"))

# 쉽게 풀은 문제


# 다른 사람 풀이, 방식은 동일하지만 코드가 마음에 들어서 긁어옴
# "Counter"사용
from collections import Counter 
def solution(want, number, discount):
    best_day = 0
    want_number = dict(zip(want, number))

    for i in range(len(discount) - 10 + 1):
        if want_number == Counter(discount[i:i + 10]):            
            best_day += 1
    return best_day

File: Programmers/Level 2/Level2_할인행사.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <queue>
#include <stack>
#include <map>
#include <set>
using namespace std;

/*
    vector<string> want         원하는 제품 목록
    vector<int> number          원하는 제품의 수량
    vector<string> discount     할인하는 제품 목록
*/
int solution(vector<string> want, vector<int> number, vector<string> discount) {
    int answer = 0;

	map<string, int> wantMap;
    for (int i = 0; i < want.size(); i++) 
        wantMap[want[i]] = number[i];
	
	for (int i = 0; i < discount.size(); i++) 
    {
        if(wantMap.find(discount[i]) != wantMap.end())
			wantMap[discount[i]]--;

        if (i >= 9) 
        {
            bool isMatch = true;
            for (const auto& item : wantMap) 
            {
                if (item.second > 0) 
                {
                    isMatch = false;
                    break;
                }
            }

            if (isMatch) 
                answer++;

            // 10일이 지난 후에는 할인하는 제품이 다시 구매 가능
            if(wantMap.find(discount[i - 9]) != wantMap.end())
                wantMap[discount[i - 9]]++;
		}
	}

    return answer;
}

// 1차 시도 : 성공
